src/sage/rings/polynomial/integer_valued_polynomials.py

```diff
@@ -769,24 +774,23 @@ def variable_shift(self, k=1):
                     S[5]
                     sage: S[5].variable_shift().variable_shift(-1)
                     S[5]
+                    sage: S[5].variable_shift(2).variable_shift(-2)
+                    S[5]
                 """
                 if k == 0:
                     return self
 
                 A = self.parent()
 
-                if k > 0:
-                    B = A.basis()
-                    resu = A.linear_combination((B[j], c) for i, c in self
-                                                for j in range(i + 1))
-                    if k == 1:
-                        return resu
-                    return resu.variable_shift(k - 1)
+                def on_basis(n):
+                    return {A._indices(j): binomial(k + n - 1 - j, n - j)
+                            for j in range(n + 1)}
 
-                resu = self - A._from_dict({i - 1: c for i, c in self if i})
-                if k == -1:
-                    return resu
-                return resu.variable_shift(k + 1)
+                from sage.data_structures.blas_dict import linear_combination
+                mc = self._monomial_coefficients
+                ret = linear_combination((on_basis(index), coeff)
+                                         for (index, coeff) in mc.items())
```

Comment:
personally, I prefer `for index, coeff in mc.items()`, and I think that this is more common in sage.

Reply:
right, indeed. Done